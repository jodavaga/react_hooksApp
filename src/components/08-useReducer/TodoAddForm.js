import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAddForm = ( { handleAddTodo } ) => {

    // custom hook to manage form, input changes
    const [ {description}, handleInputChange, reset] = useForm({
        description: ''
    });


    const handleSubmit = (e) => {
        e.preventDefault();
        // prevent add empty todo
        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: `${ description }`,
            done: false
        }

        // call function from parent that dispatch action
        handleAddTodo( newTodo );

        // reset input form, using customHook function
        reset();
    }

    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                name="description"
                className="form-control"
                onClick={ ({target}) => target.select()  }
                placeholder="Add todo..."
                autoComplete="off"
                value={ description }
                onChange={ handleInputChange }
            />

            <button
                type="submit"
                className="btn btn-primary btn-block mt-2"
            >
                Agregar
            </button>
        </form>
    )
}
