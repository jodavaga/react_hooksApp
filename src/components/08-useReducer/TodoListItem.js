import React from 'react';
import { upperCaseFirstLetter } from '../../helpers/uppercaseFirstLetter';

export const TodoListItem = ( {todo, index, handleComplete, handleDelete} ) => {
    return (
        <div 
            className="todo-item"
        >
            <p  
                className={ `${ todo.done && 'complete' }` }
                onClick={ () => handleComplete( todo.id ) }
            >
                {index+1}. { upperCaseFirstLetter(todo.desc) }
            </p>
            <button
                className="btn btn-outline-danger btn-sm"
                onClick={ () => handleDelete( todo.id ) }
            >
                Borrar
            </button>
        </div>
    )
}
