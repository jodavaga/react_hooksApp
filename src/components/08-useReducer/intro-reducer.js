// Initialstate as array
const initialState = [{
    id: 1,
    todo: 'Estudiar React',
    done: false
}]

const todoReducer = ( state = initialState, action ) => {

    // if action is agregar, then modify state with new payload
    if (action?.type === 'agregar') {
        return [...state, action.payload];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Seguir estudiando React',
    done: false
}

// action to add newTodo
const todoAddAction = {
    type: 'agregar',
    payload: newTodo
}

// if run this will add new todo to state through reducer function.
todos = todoReducer(todos, todoAddAction);

console.log(todos);
