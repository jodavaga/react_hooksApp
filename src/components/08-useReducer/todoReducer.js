export const todoReducer = (state = [], action) => {
    
    switch (action.type) {
        case 'add':
            return [...state, action.payload];

        case 'delete':
            // return and array of todos, different to deleted one
            return state.filter( todo => todo.id !== action.payload); // 1595032094606

        case 'toggle':
            return state.map( todo => 
                ( todo.id === action.payload )
                    ? { ...todo, done: !todo.done }
                    : todo    
            );
            
        // same as above
        case 'toggle-old':
            return state.map( todo => {
                if ( todo.id === action.payload ) {
                    // update only one prop on the todo obj
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }

                return todo;
            })
    
        default:
            return state;
            break;
    }
}