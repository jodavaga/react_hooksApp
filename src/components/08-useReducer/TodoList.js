import React from 'react';
import PropTypes from 'prop-types';

import { TodoListItem } from './TodoListItem';

export const TodoList = ( {todos, handleComplete, handleDelete} ) => {
    return (
        <>
            { todos.length > 0 &&
                <ul className="todo-list">
                    {
                        todos.map( (item, i) => {
                            return (
                                <TodoListItem 
                                    key={ item.id }
                                    todo={ item }
                                    index={ i }
                                    handleComplete={ handleComplete }                                                                                
                                    handleDelete={ handleDelete }                                    
                                />
                            )
                        })    
                    }
                </ul>
            }
        </>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleComplete: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}
