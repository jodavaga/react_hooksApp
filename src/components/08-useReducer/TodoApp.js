import React, { useReducer } from 'react';
import { todoReducer } from './todoReducer';


const initialState = [{
    id: new Date().getTime(),
    desc: 'Estudiar React',
    done: false
}];

export const TodoApp = () => {

    const [ todos ] = useReducer(todoReducer, initialState);

    return (
        <div className="container">
            <h1>Todo APP</h1>
            <hr />

            <ul>
                {
                    todos.map( ({id, desc}) => {
                        return (
                            <li key={ id }> { desc} </li>
                        )
                    })    
                }
            </ul>
        </div>
    )
}
