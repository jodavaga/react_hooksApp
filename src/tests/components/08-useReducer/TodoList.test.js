import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('test on <TodoList />', () => {

    const handleDelete = jest.fn();
    const handleComplete = jest.fn();
    
    const wrapper = shallow( 
        <TodoList 
            todos={ demoTodos }
            handleDelete={ handleDelete }
            handleComplete={ handleComplete }
        /> 
    );

    test('should render component', () => {
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('should have demoTodos.length amount of <TodoListItem />', () => {
        
        const todoItems = wrapper.find('TodoListItem');

        expect( todoItems.length ).toBe( demoTodos.length );
        // console.log( todoItems.at(0).props() )
        expect( todoItems.at(0).prop('handleDelete') ).toEqual( expect.any( Function ) );

    });
    
    test('should not render TodoListItems if not todos', () => {
        
        const wrapper = shallow( 
            <TodoList 
                todos={ [] }
                handleDelete={ handleDelete }
                handleComplete={ handleComplete }
            /> 
        );

        expect( wrapper.exists('.todo-list') ).toBe(false);
    });
    
})
