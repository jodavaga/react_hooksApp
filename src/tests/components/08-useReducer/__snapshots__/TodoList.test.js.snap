// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on <TodoList /> should render component 1`] = `
<Fragment>
  <ul
    className="todo-list"
  >
    <TodoListItem
      handleComplete={[MockFunction]}
      handleDelete={[MockFunction]}
      index={0}
      key="1"
      todo={
        Object {
          "desc": "Estudiar React",
          "done": false,
          "id": 1,
        }
      }
    />
    <TodoListItem
      handleComplete={[MockFunction]}
      handleDelete={[MockFunction]}
      index={1}
      key="2"
      todo={
        Object {
          "desc": "Estudiar React",
          "done": false,
          "id": 2,
        }
      }
    />
  </ul>
</Fragment>
`;
